src/ucp/proto/proto_perf.c

```diff
@@ -0,0 +1,400 @@
+/**
+ * Copyright (c) NVIDIA CORPORATION & AFFILIATES, 2024. ALL RIGHTS RESERVED.
+ *
+ * See file LICENSE for terms.
+ */
+
+#ifdef HAVE_CONFIG_H
+#include "config.h"
+#endif
+
+#include "proto_perf.h"
+#include "proto_debug.h"
+
+#include <ucs/datastruct/list.h>
+#include <ucs/debug/memtrack_int.h>
+#include <ucs/sys/string.h>
+
+
+struct ucp_proto_perf_segment {
+    /* List element */
+    ucs_list_link_t       list;
+
+    /* Start value of this segment (inclusive) */
+    size_t                start;
+
+    /* End value of this segment (inclusive) */
+    size_t                end;
+
+    /* Associacted performance node */
+    ucp_proto_perf_node_t *node;
+
+    /* Linear function representing value of each contributing factor */
+    ucs_linear_func_t     factors[UCP_PROTO_PERF_FACTOR_LAST];
+};
+
+/*
+ * Protocol performance structure.
+ */
+struct ucp_proto_perf {
+    /* Name of the protocol */
+    char            name[UCP_PROTO_DESC_STR_MAX];
+
+    /* List of segments */
+    ucs_list_link_t segments;
+};
+
+/*
+ * Iterate over the segments of the protocol perf structure function.
+ */
+#define ucp_proto_perf_segment_foreach(_perf, _seg) \
+    ucs_list_for_each((_seg), &(_perf)->segments, list)
+
+static const char *ucp_proto_perf_factor_names[] = {
+    [UCP_PROTO_PERF_FACTOR_LOCAL_CPU]  = "cpu",
+    [UCP_PROTO_PERF_FACTOR_REMOTE_CPU] = "rem_cpu",
+    [UCP_PROTO_PERF_FACTOR_TL_SEND]    = "send",
+    [UCP_PROTO_PERF_FACTOR_TL_RECEIVE] = "recv",
+    [UCP_PROTO_PERF_FACTOR_LATENCY]    = "lat",
+    [UCP_PROTO_PERF_FACTOR_SINGLE]     = "sngl",
+    [UCP_PROTO_PERF_FACTOR_MULTI]      = "mult",
+    [UCP_PROTO_PERF_FACTOR_LAST]       = NULL
+};
+
+static void ucp_proto_perf_check(ucp_proto_perf_t *perf)
+{
+#if ENABLE_ASSERT
+    ucp_proto_perf_segment_t *seg;
+    size_t min_start;
+
+    min_start = 0;
+    ucp_proto_perf_segment_foreach(perf, seg) {
+        ucs_assertv((seg->start >= min_start) && (seg->start <= seg->end),
+                    "perf=%p seg->start=%zu seg->end=%zu min_start=%zu", perf,
+                    seg->start, seg->end, min_start);
+        min_start = seg->end + 1;
```

Comment:
do we have a case where `seg->end` is UINT64_MAX? if yes we might want to check that when min_start is lower than seg->end, we are the last segment?